datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssigner.java

```diff
@@ -85,24 +93,31 @@
     });
 
     // build a priority queue of tasks based on throughput
-    List<String> tasks = new ArrayList<>(newPartitions.keySet());
+    // only add tasks that can accommodate more partitions in the queue
+    List<String> tasks = newPartitions.keySet().stream()
+        .filter(t -> newPartitions.get(t).size() < maxPartitionsPerTask)
+        .collect(Collectors.toList());
     PriorityQueue<String> taskQueue = new PriorityQueue<>(Comparator.comparing(taskThroughputMap::get));
     taskQueue.addAll(tasks);
 
     // assign partitions with throughput info one by one, by putting the heaviest partition in the lightest task
-    while (recognizedPartitions.size() > 0) {
+    while (recognizedPartitions.size() > 0 && taskQueue.size() > 0) {
        String heaviestPartition = recognizedPartitions.remove(recognizedPartitions.size() - 1);
        int heaviestPartitionThroughput = partitionInfoMap.get(heaviestPartition).getBytesInKBRate();
        String lightestTask = taskQueue.poll();
        newPartitions.get(lightestTask).add(heaviestPartition);
        taskThroughputMap.put(lightestTask, taskThroughputMap.get(lightestTask) + heaviestPartitionThroughput);
-       taskQueue.add(lightestTask);
+       int currentNumPartitions = newPartitions.get(lightestTask).size();
+       // don't put the task back in the queue if the number of its partitions is maxed out
+       if (currentNumPartitions < maxPartitionsPerTask) {
+         taskQueue.add(lightestTask);
+       }
     }
 
-    // TODO implement a mechanism to prevent tasks from having more than partitionsPerTask partitions
     // assign unrecognized partitions with round-robin
     int index = 0;
     for (String partition : unrecognizedPartitions) {
+      index = findTaskWithRoomForAPartition(tasks, newPartitions, index, maxPartitionsPerTask);
       String currentTask = tasks.get(index);
       newPartitions.get(currentTask).add(partition);
       index = (index + 1) % tasks.size();
```

Comment:
It's not a no-op. `findTaskWithRoomForAPartition` is returning the index of the first available task, starting from the startIndex. If I don't increment the index, it will keep returning the same index until the task with the start index is exhausted.